### PR TITLE
Elevator subsystem

### LINKED ISSUE #23: Elevator Subsystem
This subsystem will need to include:

- Encoders for elevator status

Possible but likely functions include:

- Move elevator (to move elevator to our goal)
- Elevator rotation (to get the elevator's rotation)

We will also need to know these different positions either using a REV Through Bore Encoder or canCoder. These positions include:

- Pole 2 / Block State 2 (When attempting to place cone / cube in the top section of the grid)
- Pole 1 / Block State 1 (When attempting to place cone / cube in the middle section of the grid)
- Ground (When resting out and ready to grab objects off of floor)
- Stored (When pushed inside of robot before match start)

Will need PID. https://docs.wpilib.org/en/stable/docs/software/advanced-controls/controllers/pidcontroller.html
REV API Docs: https://codedocs.revrobotics.com/java/allclasses.html

### PR DESCRIPTION
**Pull request by gvaldez7206**
Closes #23

### CODE REVIEW

File: src/main/java/frc/robot/commands/MoveElevator.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import frc.robot.Constants;
+import frc.robot.subsystems.Elevator;
+
+/**
+ * Creates a command for movine the elevator.
+ */
+public class MoveElevator extends CommandBase {
+
+    private CANSparkMax motor1 = Elevator.leftElevatorMotor;
+    private CANSparkMax motor2 = Elevator.rightElevatorMotor;
```
Comment (legoguy1000) [inline src/main/java/frc/robot/commands/MoveElevator.java:15]: This should only be defined within subsystem

File: src/main/java/frc/robot/commands/MoveElevator.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    private DutyCycleEncoder dutyCycleEncoder = Elevator.elevatorEncoder;
```
Comment (legoguy1000) [inline src/main/java/frc/robot/commands/MoveElevator.java:16]: This should only be defined within subsystem

File: src/main/java/frc/robot/commands/MoveElevator.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    private double encoderGoal;
+
+    /**
+     * Creates a command for movine the elevator.
+     *
+     * @param rotation Current rotation value to determine where the motor is.
+     */
+    public MoveElevator(double rotation) {
+        this.encoderGoal = rotation;
```
Comment (legoguy1000) [inline src/main/java/frc/robot/commands/MoveElevator.java:25]: Missing subsystem argument.
Missing addRequirements()

File: src/main/java/frc/robot/commands/MoveElevator.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    @Override
+    public void execute() {
+        Elevator.runMotor(encVal(), encoderGoal);
```
Comment (legoguy1000) [inline src/main/java/frc/robot/commands/MoveElevator.java:30]: should be a class variable from constructor

File: src/main/java/frc/robot/commands/MoveElevator.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    @Override
+    public boolean isFinished() {
+        return (this.encoderGoal == 0);
+    }
+
+    public double encVal() {
+        return (dutyCycleEncoder.getAbsolutePosition() - Constants.Elevator.encoderOffSet);
+    }
```
Comment (legoguy1000) [inline src/main/java/frc/robot/commands/MoveElevator.java:47]: Should be in subsystem

Comment (legoguy1000) [reply src/main/java/frc/robot/commands/MoveElevator.java:47]: no () needed

File: src/main/java/frc/robot/subsystems/Elevator.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    public Elevator() {
+
+    }
+
+    /**
+     * Creates runMotor method to move the motors.
+     */
+    public static void runMotor(double currentVal, double goal) {
+        leftElevatorMotor.set(elevatorPID.calculate(currentVal, goal));
+        rightElevatorMotor.set(elevatorPID.calculate(currentVal, goal));
```
Comment (legoguy1000) [inline src/main/java/frc/robot/subsystems/Elevator.java:44]: only 1 motor per Mr Rucinski

File: src/main/java/frc/robot/subsystems/Elevator.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * Creates subsystem and functions for the elevator.
+ */
+public class Elevator extends SubsystemBase {
+
+    public static final CANSparkMax leftElevatorMotor =
+        new CANSparkMax(Constants.Elevator.leftElevatorMotorID, MotorType.kBrushless);
+    public static final CANSparkMax rightElevatorMotor =
+        new CANSparkMax(Constants.Elevator.rightElevatorMotorID, MotorType.kBrushless);
```
Comment (legoguy1000) [inline src/main/java/frc/robot/subsystems/Elevator.java:20]: only 1 motor per Mr Rucinski

File: src/main/java/frc/robot/subsystems/Elevator.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+    public static final DutyCycleEncoder elevatorEncoder = new DutyCycleEncoder(99);
+
+    // we'll use it later
+    private final ElevatorFeedforward elevatorFF = new ElevatorFeedforward(
+        Constants.Elevator.elevatorKSVolts, Constants.Elevator.elevatorKGVolts,
+        Constants.Elevator.elevatorKVVoltsSecondsPerRotation);
```
Comment (legoguy1000) [inline src/main/java/frc/robot/subsystems/Elevator.java:27]: not using feed forward

File: src/main/java/frc/robot/subsystems/Elevator.java (hunk 7 above)
Comment (legoguy1000) [inline src/main/java/frc/robot/subsystems/Elevator.java:20]: Should be private

File: src/main/java/frc/robot/subsystems/Elevator.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+ * Creates subsystem and functions for the elevator.
+ */
+public class Elevator extends SubsystemBase {
+
+    public static final CANSparkMax leftElevatorMotor =
+        new CANSparkMax(Constants.Elevator.leftElevatorMotorID, MotorType.kBrushless);
+    public static final CANSparkMax rightElevatorMotor =
+        new CANSparkMax(Constants.Elevator.rightElevatorMotorID, MotorType.kBrushless);
+
+    public static final DutyCycleEncoder elevatorEncoder = new DutyCycleEncoder(99);
```
Comment (legoguy1000) [inline src/main/java/frc/robot/subsystems/Elevator.java:22]: should be private

File: src/main/java/frc/robot/commands/MoveElevator.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    @Override
+    public void end(boolean interrupted) {
+        motor1.set(0);
+        motor2.set(0);
+    }
+
+    @Override
+    public boolean isFinished() {
+        return (this.encoderGoal == 0);
+    }
```
Comment (legoguy1000) [inline src/main/java/frc/robot/commands/MoveElevator.java:43]: This doesn't make sense, when does encoderGoal ever change from initial set? Will this ever trigger? no ()

File: src/main/java/frc/robot/commands/MoveElevator.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    @Override
+    public void execute() {
+        Elevator.runMotor(encVal(), encoderGoal);
+
+    }
+
+    @Override
+    public void end(boolean interrupted) {
+        motor1.set(0);
+        motor2.set(0);
```
Comment (legoguy1000) [inline src/main/java/frc/robot/commands/MoveElevator.java:37]: 1 motor

File: src/main/java/frc/robot/commands/MoveElevator.java (hunk 1 above)
Comment (legoguy1000) [inline src/main/java/frc/robot/commands/MoveElevator.java:15]: only 1 motor

**Review by legoguy1000 — CHANGES_REQUESTED** [review]

**Review by owenflatman — APPROVED** [review]
Looks good 👍

**Review by legoguy1000 — APPROVED** [review]

**Pull request merged**